addons/base/views.py

```diff
@@ -281,100 +263,124 @@ def get_auth(auth, **kwargs):
-                                        version=fileversion.identifier if fileversion else None,
-                                        path=path,
-                                    )
-                                except es_exceptions.ConnectionError:
-                                    log_exception()
-        if fileversion and provider_settings:
-            region = fileversion.region
-            credentials = region.waterbutler_credentials
-            waterbutler_settings = fileversion.serialize_waterbutler_settings(
-                node_id=provider_settings.owner._id,
-                root_id=provider_settings.root_node._id,
-            )
-    # If they haven't been set by version region, use the NodeSettings or Preprint directly
-    if not (credentials and waterbutler_settings):
-        credentials = node.serialize_waterbutler_credentials(provider_name)
-        waterbutler_settings = node.serialize_waterbutler_settings(provider_name)
 
+@collect_auth
+def get_auth(auth, **kwargs):
+    waterbutler_data = decrypt_and_decode_jwt_payload()
+    node = get_authenticated_node(waterbutler_data['nid'])
+    action = waterbutler_data['action']
+
+    # Fallback to cookie-based/token authentication if the user is not authenticated yet
+    if not auth.user:
+        if 'cookie' in waterbutler_data:
+            user = OSFUser.from_cookie(waterbutler_data.get('cookie'))
+            if user:
+                auth.user = user
+            else:
+                raise HTTPError(http_status.HTTP_401_UNAUTHORIZED)
+
+        authorization = request.headers.get('Authorization')
+        if authorization and authorization.startswith('Bearer '):
+            auth.user = authenticate_via_oauth_bearer_token(node, action)
+
+    action = waterbutler_data['action']
+    if not check_node_permission(node, auth, action):
+        raise HTTPError(http_status.HTTP_401_UNAUTHORIZED)
+
+    if not isinstance(node, Preprint):
+        provider = node.get_addon(waterbutler_data['provider'])
+    else:
+        provider = None
+
+    if not provider and not isinstance(node, Preprint):
+        raise HTTPError(http_status.HTTP_400_BAD_REQUEST)
+    elif waterbutler_data['provider'] == 'osfstorage':
+        fileversion = get_file_version_from_wb(waterbutler_data)
+        credentials = fileversion.region.waterbutler_credentials
+        waterbutler_settings = fileversion.serialize_waterbutler_settings(
+            node_id=provider.owner._id,
+            root_id=provider.root_node._id,
+        )
+        if action == 'render':
+            file_signals.file_viewed.send(auth=auth, fileversion=fileversion)
+        if action == 'download':
+            file_signals.file_downloaded.send(auth=auth, fileversion=fileversion)
+    else:
+        credentials = node.serialize_waterbutler_credentials(waterbutler_data['provider'])
+        waterbutler_settings = node.serialize_waterbutler_settings(waterbutler_data['provider'])
```

Comment:
Cleanest approach here (especially for GV integration) is probably to reduce this to
```
elif ... `osfstorage`:
  credentials, settings = get_osf_storage_settings(. . .)
else:
   credentials, settings = get_addon_wb_config(. . .)
```
and then the `get_addon_wb_config` can be where the flag is used to fetch stuffs from GV